diff --git a/bbc.c b/bbc.c
index b8e3bb7a..d6e22925 100644
--- a/bbc.c
+++ b/bbc.c
@@ -1171,7 +1171,8 @@ bbc_framebuffer_ready_callback(void* p,
   struct bbc_message message;
 
   struct bbc_struct* p_bbc = (struct bbc_struct*) p;
-  int do_wait_for_render = !p_bbc->fast_flag;
+  int do_wait_for_render = !p_bbc->fast_flag ||
+                           video_has_paint_timer_triggered(p_bbc->p_video);
 
   message.data[0] = k_message_vsync;
   message.data[1] = do_full_render;
diff --git a/video.c b/video.c
index a2707041..0dc3fd2e 100644
--- a/video.c
+++ b/video.c
@@ -2185,4 +2185,9 @@ video_get_crtc_state(struct video_struct* p_video,
   *p_is_in_dummy_raster = p_video->in_dummy_raster;
 }
 
+int
+video_has_paint_timer_triggered(struct video_struct* p_video) {
+  return p_video->has_paint_timer_triggered;
+}
+
 #include "test-video.c"
diff --git a/video.h b/video.h
index 0219e2c6..9af23d18 100644
--- a/video.h
+++ b/video.h
@@ -74,4 +74,6 @@ void video_get_crtc_state(struct video_struct* p_video,
                           int* p_is_in_vert_adjust,
                           int* p_is_in_dummy_raster);
 
+int video_has_paint_timer_triggered(struct video_struct* p_video);
+
 #endif /* BEEBJIT_VIDEO_H */
